doc/read-the-docs-site/adr/0005-plutus-tx-compiler-and-library.md

```diff
@@ -0,0 +1,250 @@
+
+This is the reason why let-bindings have been non-strict by default in Plutus Tx, and creating strict let-bindings requires using bang patterns.
+
+We have a few options:
+
+1. Do nothing, and encourage people to use the `Strict` GHC extension, which makes     let-bindings strict by default.
+   This is not technically "strict by default", since the `Strict` extension is off by default, but it is still ergonomically better than using bang patterns on every binding.
+
+   One argument against this is that a module may contain both on-chain and off-chain code, some modules contain off-chain code only, and the same piece of code may be both on-chain and off-chain.
+   For off-chain code one typically doesn't want to enable `Strict`.
+
+   Another disadvantage of using `Strict` is that it may cause GHC to do additional optimizations in ways that can break the Plutus Tx compiler.
+   As a result, in addition to `Strict`, one may need to also use one or more of the following GHC flags: `-fno-strictness`, `-fno-specialise`, `-fno-spec-constr`, `-fno-unbox-strict-fields`, `-fno-unbox-small-strict-fields`.
+   Not only is it inconvenient, but failing to do so may lead to obscure compilation errors.
+2. Enable `Strict` automatically.
+   This can be done in a GHC driver plugin.
+   But like said in Option 1, one doesn't usually want to enable `Strict` on off-chain code.
+   Thus at the very least, we'd need to provide the option to turn it off via `NoStrict`, and this is tricky to implement.
+3. Introduce the following function:
+
+  ```haskell
+  nonstrict :: forall a. a -> a
+  nonstrict = id
+  {-# NOINLINE nonstrict #-}
+  ```
+
+  Then, to create a non-strict let-binding, we write
+
+  ```haskell
+  let x = nonstrict rhs in body
+  ```
+
+  Because of the NOINLINE pragma, `nonstrict` is guaranteed to be visible to the Plutus Tx compiler.
+  Whenever the Plutus Tx compiler sees `let x = nonstrict rhs`, it first compiles `rhs` normally; then it creates a nonstrict binding in PIR: `let ~x = rhs_compiled`.
+
+The advantages of making let-bindings strict by default, and using `nonstrict` for non-strict bindings are:
+
+- It makes it more convenient, rather than less convenient, to create strict bindings than non-strict bindings, which is what a strict language should do.
+- It works for both on-chain code and off-chain code. A let-binding is strict by default when used in on-chain code, and non-strict by default otherwise.
+
+The disadvantages are:
+
+- It is a magic function that needs special support in the Plutus Tx compiler.
+- This is a critical one: it is susceptible to GHC inlining bindings unconditionally.
+  If GHC decides to turn `let x = rhs in ...x...` into `...rhs...`, then `x` effectively becomes non-strict, and there's nothing the Plutus Tx compiler can do to recover `x`.
+
+If we can modify the Plutus Tx compiler such that we can turn off GHC's pre-inliner (which will be discussed later), then option 3 becomes the most preferrable.
```

Comment:
I think this points to an under-appreciated benefit of option 1: if we can persuade GHC that the source language is strict, then we should have to worry a _lot less_ in general about GHC mis-optimizing things, because GHC should have a compatible view of the world to us.